plugins/ipfs_cli_command/ipfs_cli_command/ipfs_utils.py

```diff
@@ -0,0 +1,166 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""Ipfs utils for `ipfs cli command`."""
+import os
+import shutil
+import signal
+import subprocess
+import time
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import ipfshttpclient
+
+
+class IPFSDaemon:
+    """
+    Set up the IPFS daemon.
+
+    :raises Exception: if IPFS is not installed.
+    """
+
+    def __init__(self, timeout: float = 15.0):
+        """Initialise IPFS daemon."""
+        # check we have ipfs
+        self.timeout = timeout
+        res = shutil.which("ipfs")
+        if res is None:
+            raise Exception("Please install IPFS first!")
+        process = subprocess.Popen(  # nosec
+            ["ipfs", "--version"], stdout=subprocess.PIPE, env=os.environ.copy(),
+        )
+        output, _ = process.communicate()
+        if b"0.6.0" not in output:
+            raise Exception(
+                "Please ensure you have version 0.6.0 of IPFS daemon installed."
+            )
+        self.process = None  # type: Optional[subprocess.Popen]
+
+    def __enter__(self) -> None:
+        """Run the ipfs daemon."""
+        self.process = subprocess.Popen(  # nosec
+            ["ipfs", "daemon"], stdout=subprocess.PIPE, env=os.environ.copy(),
+        )
+        print("Waiting for {} seconds the IPFS daemon to be up.".format(self.timeout))
+        time.sleep(self.timeout)
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:  # type: ignore
+        """Terminate the ipfs daemon."""
+        if self.process is None:
+            return
+        self.process.send_signal(signal.SIGTERM)
+        self.process.wait(timeout=30)
+        poll = self.process.poll()
+        if poll is None:
+            self.process.terminate()
+            self.process.wait(2)
+
+
+class RemoveError(Exception):
+    """Exception on remove."""
+
+
+class PublishError(Exception):
+    """Exception on publish."""
+
+
+class NodeError(Exception):
+    """Exception for node connection check."""
+
+
+class IPFSTool:
+    """IPFS tool to add, publish, remove, download directories."""
+
+    def __init__(self, client_options: Optional[Dict] = None):
+        """
+        Init tool.
+
+        :param client_options: dict, options for ipfshttpclient instance.
+        """
+        self.client = ipfshttpclient.Client(**(client_options or {}))
+
+    def add(self, dir_path: str) -> Tuple[str, str, List]:
+        """
+        Add directory to ipfs.
+
+        It wraps into directory.
+
+        :param dir_path: str, path to dir to publush
+
+        :return: dir name published, hash, list of items processed
+        """
+        response = self.client.add(dir_path, recursive=True, wrap_with_directory=True)
```

Comment:
what about the pin flag?

Reply:
it's true by default, can set explicitly

Reply:
ok, but we should make this false by default and then allow a flag to be passed to make it true.

Reply:
done